src/execution/periodic_epoch.rs

```diff
@@ -70,6 +71,56 @@ impl PeriodicEpochConfig {
     }
 }
 
+// Struct used to make it clear what we are doing with tracing spans,
+// for internal use of this module.
+struct TracingHelper {
```

Comment:
I think it'd be useful to try out a different span hierarchy, and think of one more useful to people using dataflows. Here are three ideas that come to mind:

## 1. Data-centric

In a sense, items flowing through a dataflow are the similar to HTTP requests flowing through a microservice graph. If you're writing a dataflow and want to see what your data is doing, you'll be thinking about how to individual items flow. So do something like assign a "request ID" to each item emitted from the input so the item is the top-most, then have each operator's work on that item contribute a sub-span. End the top-level item span when all the resulting items are sent through output.

I'm not totally sure how stateful operators or windowing or aggregation would work in this context. Like if you're summing values for an hour long window, do you keep each item open until the sum output is complete? We'd then need to allow the state items to build up sets of relevant request IDs. I don't think there's any analogous problem in the microservices world as requests never coalesce?

This seems complicated to get right and revolves having a lot of metadata attached to each `TdPyAny` flowing through, but would be super cool.

One downside is you don't see operational interactions: like if one operator is hogging all the CPU time, that might be hard to know. But maybe it doesn't matter?

## 2. Epoch-centric

What I mentioned in another comment, but again: each top-level span is an epoch and you have sub-spans the part of each operator activation that works on each specific epoch. This would let you see interleaving of work for different epochs and know if there's some internal bottlenecks. But I'm not sure if it's really that helpful. It doesn't reveal anything about data dependencies.

## 3. Operational Execution-centric

This feels like a mode for us as Bytewax developers. Have each call to `Worker.step` be a top-level span, and see what operator's awakenings take up fractions of that time. We'd get a good sense of where each worker is spending CPU or wait time, but you don't learn much about the data directly as each batch is monolithic. You learn about the efficiency of relative parts and internal operator design.

I think the biggest benefit of this third option is that it requires no modifications to the dataflow itself to thread through tracing information.

Reply:
I pushed the changes I was working on on Friday. It is a mix of 3 and a different interpretation of 2.

Basically instead of using actual epochs to subdivide the spans, I added a more agnostic `periodic` span, which is closed and reopened every 10 seconds (we could make this configurable).

As I said in another comment, we could now have a proper `step` sudivision, since we handle spans where we call `worker.step`.

The problem is that since we never park the worker, we get a lot of empty `step` spans (like 10x the number of spans without it on the wikistream example). I'm going to check if I can filter out or aggregate empty step spans with a custom sampler anyway, because I find the `step` subdivision can be useful, but not if it floods the data with "useless" info.

Ideally I'd love to have `1`, a data centric subdivision, but as you noted it is non trivial as soon as you introduce stateful operators and/or multiple captures and event time processing (since spans are grouped together by system time).

We could handle that by doing something similar to what we do for recovery, and observes the data flowing through and build spans manually with the proper info for correct grouping, but this will add an overhead.

I think tracing itself should be as low cost as possible, but in the context of observability this is something I'd like to have.

I think it makes sense to have tracing itself be operational centric, but still give the users another way to observe the data flowing, which might be non zero-cost, so that if you want more advanced views of your data you can opt in knowing it doesn't come for free.